docs/RFCs/0004-crashtracker-design.md

```diff
@@ -0,0 +1,146 @@
+# RFC 0004: Crashtracker design
+
+## Background
+
+A crashtracker implementation must balance between two conflicting aims.
+On the one hand, its purpose is to **reliably collect and report the information necessary to diagnose and debug all crashes** in the tracked process.
+On the other hand, a crashtracker, like other instrumentation technology, should ideally **cause no observable difference in the execution of the processes where it is installed**.
+These aims are in tension: design decisions which increase the quality and reliability of data-collection often increase the (potential) for impact on the customer system.
+
+This document lays out the design priorities of the crashtracker, describes how the current architecture affects/implements those priorities, and then lays out a plan for future improvements to the crashtracking ecosystem.
+
+## Design Priorities
+
+1.  When the system is functioning normally, the crashtracker should, to the greatest extent possible, **have no impact on the functioning of a non-crashing system**.
+2.  When a crash does occur, the crashtracker should **prioritize the reliable collection and reporting of crash information**.
+3.  When a crash does occur, the crashtracker should **make a best-effort attempt to minimize impact** on the crashing process.
+
+## Architectural model
+
+### Collector
+
+For all languages other than .Net and Java, this is a signal handler.
+The collector interacts with the system in the following ways:
+
+1.  Registers a signal handler (`sigaction`).
+    The old handler is saved into a global variable to enable changing.
+    - Risks to normal operation
+      - Chaining handlers is not atomic.
+        It is possible for concurrent races to occur.
+        Mitigation: doing this operation once, early during system initialization.
+      - If the system expects and handles `sigsegv`, `sigbus` and `sigabort` as part of its ordinary functioning, this can violate principle 1.
+        This is not common for most languages, but Java and .Net turn some crashes into recoverable user exceptions.
```

Comment:
I believe this is not the reason for Java; afaik Java (or more specifically the hotspot VM) is using SEGVs for internal VM mechanisms (e.g. GC optimizations and the like).

Not that it matters for this doc! But I would suggest being a bit more generic here to avoid being slightly incorrect -- e.g. stating that these VMs require direct control of these signals for implementation reasons and that we can't touch them there.

Reply:
fixed